Make sure nested worker get controlled if matching a service worker registration

youennf opened the pull request:
#### 056c2a9cad9c31deb641fcb7ca0d8e1291d92e8c
<pre>
Make sure nested worker get controlled if matching a service worker registration
<a href="https://bugs.webkit.org/show_bug.cgi?id=247619">https://bugs.webkit.org/show_bug.cgi?id=247619</a>
rdar://problem/102090425

Reviewed by Chris Dumez.

Allow to register WorkerScriptLoader living of the main thread (nested workers do create them).
Instead of registering a WorkerScriptLoader*, we now register a callback which is called when we need to set a worker as controlled.
Covered by unskipped test.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/nested-blob-url-workers.https-expected.txt: Added.
* Source/WebCore/workers/WorkerScriptLoader.cpp:
(WebCore::workerScriptLoaderControlledCallbackMap):
(WebCore::accessWorkerScriptLoaderMap):
(WebCore::WorkerScriptLoader::~WorkerScriptLoader):
(WebCore::WorkerScriptLoader::loadAsynchronously):
(WebCore::WorkerScriptLoader::getWorkerClientControlledCallback):
(): Deleted.
(WebCore::WorkerScriptLoader::fromScriptExecutionContextIdentifier): Deleted.
* Source/WebCore/workers/WorkerScriptLoader.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::setServiceWorkerClientIsControlled):

Canonical link: <a href="https://commits.webkit.org/256637@main">https://commits.webkit.org/256637@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42e98acf6b0acf73c37aed6be1d829e688aabc46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96071 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29114 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105622 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165950 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100049 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5436 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34080 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101439 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4015 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82676 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31047 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73880 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39812 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19302 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37487 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20645 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4606 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42083 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43250 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39908 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/workers/WorkerScriptLoader.cpp:
```diff
@@ -363,20 +365,50 @@ WorkerFetchResult WorkerScriptLoader::fetchResult() const
+{
+    Locker lock(m_activeServiceWorkerDataLock);
+    std::optional<ServiceWorkerData> data;
+    std::exchange(data, m_activeServiceWorkerData);
```
I think this would be clearer:
`std::optional<ServiceWorkerData> data = std::exchange(m_activeServiceWorkerData, std::nullopt);`

In your current code, it is not obvious that m_activeServiceWorkerData is getting cleared.

youennf replied:
Done

cdumez commented on Source/WebCore/workers/WorkerScriptLoader.h:
```diff
@@ -98,11 +98,29 @@ class WorkerScriptLoader : public RefCounted<WorkerScriptLoader>, public Threada
+
+        ScriptExecutionContextIdentifier m_clientIdentifier;
+        Lock m_activeServiceWorkerDataLock;
+        std::optional<ServiceWorkerData> m_activeServiceWorkerData;
```
Please use WTF_GUARDED_BY_LOCK().

youennf replied:
Done

cdumez commented on Source/WebCore/workers/WorkerScriptLoader.h:
```diff
@@ -98,11 +98,29 @@ class WorkerScriptLoader : public RefCounted<WorkerScriptLoader>, public Threada
+
+    void setControllingServiceWorker(ServiceWorkerData&&);
+    std::optional<ServiceWorkerData> takeServiceWorkerData();
+    WEBCORE_EXPORT static RefPtr<ServiceWorkerDataManager> getServiceWorkerDataManager(ScriptExecutionContextIdentifier);
```
We should use a "get" prefix.

youennf replied:
Changed the name.

youennf committed:
Make sure nested worker get controlled if matching a service worker registration
https://bugs.webkit.org/show_bug.cgi?id=247619
rdar://problem/102090425

Reviewed by Chris Dumez.

Allow to register WorkerScriptLoader living of the main thread (nested workers do create them).
Instead of registering a WorkerScriptLoader*, we now register a callback which is called when we need to set a worker as controlled.
Covered by unskipped test.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/nested-blob-url-workers.https-expected.txt: Added.
* Source/WebCore/workers/WorkerScriptLoader.cpp:
(WebCore::workerScriptLoaderControlledCallbackMap):
(WebCore::accessWorkerScriptLoaderMap):
(WebCore::WorkerScriptLoader::~WorkerScriptLoader):
(WebCore::WorkerScriptLoader::loadAsynchronously):
(WebCore::WorkerScriptLoader::getWorkerClientControlledCallback):
(): Deleted.
(WebCore::WorkerScriptLoader::fromScriptExecutionContextIdentifier): Deleted.
* Source/WebCore/workers/WorkerScriptLoader.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::setServiceWorkerClientIsControlled):

Canonical link: https://commits.webkit.org/256637@main
